CSI Manila driver

gman0 opened the pull request:
**What this PR does / why we need it**:
Adds CSI Manila driver, currently supports only CephFS and NFS. ~~I've only tested with CephFS so far.~~ Tested with CephFS and NFS.

**Which issue this PR fixes**: fixes #394 

### Design considerations / a brief description of what does what

#### "proxy" architecture

This CSI driver acts as a provisioner of Manila shares, all node-related operations (attachments, mounts) are carried out by another CSI driver, dedicated for that particular filesystem. This is achieved by forwarding CSI RPCs to the other CSI driver.

For example, creating and mounting a CephFS Manila share would go like this:

CO requests a volume provisioning with `CreateVolume`:
```
                        +--------------------+
                        |       MASTER       |
+------+  CreateVolume  |  +--------------+  |
|  CO  +------------------>+  csi-manila  |  |
+------+                |  +--------------+  |
                        +--------------------+

1. Authenticate with Manila v2 client
2. Create a CephFS share
3. Create a cephx access rule
```

And when CO publishes the volume onto a node with `NodePublishVolume`:
```
                             +--------------------------------------+
                             |                 NODE                 |
+------+  NodePublishVolume  |  +--------------+                    |
|  CO  +----------------------->+  csi-manila  |                    |
+------+                     |  +------+-------+                    |
                             |         |                            |
                             |         | FORWARD NodePublishVolume  |
                             |         V                            |
                             |  +------+-------+                    |
                             |  |  csi-cephfs  |                    |
                             |  +--------------+                    |
                             +--------------------------------------+

1. Authenticate with Manila v2 client
2. Retrieve the CephFS share
3. Retrieve the cephx access rule
4. Connect to csi-cephfs socket
5. Call csi-cephfs's NodePublishVolume, return its response
```

So the intention here is to deal with only Manila side of things, and leave all filesystem specifics to another CSI driver that's specialized for that particular fs.

If connection with the proxy'd CSI driver cannot be established, csi-manila will return an error and will let the CO retry.

Downsides:

From the documentation
> A single instance of the driver may serve only a single Manila share protocol. To support multiple share protocols, multiple deployments of the driver need to be made. In order to avoid deployment collisions, each instance of the driver should be named differently, e.g. `csi-manila-cephfs`, `csi-manila-nfs`.

The initial idea was to encompass all Manila share protocols within a single instance of csi-manila. Due to limitations of CSI, this cannot be achieved and there will have to be multiple instances of csi-manila running in order to support multiple share protocols, one for each. There are couple of RPCs in Node Service that don't bring enough context to decide which proxy'd driver that particular RPC should be forwarded to (e.g. `NodeGetCapabilities` would be ambiguous and we can't know which proxy'd driver should answer) => therefore with current design, there's only a single proxy'd driver per csi-manila instance. Such "multiplexer" design is out-of-scope for CSI anyway IMHO.

#### Concurrency

`CreateVolume` is so far the most time-expensive operation in the driver because it waits till the share is in _Available_ state, and for CephFS shares this waiting is even more pronounced because it waits till the Ceph key for the share is available. This means that the CO may retry the `CreateVolume` RPC for the same volume many times before it's actually done creating.

In Kubernetes, this means there will be multiple `CreateVolume` RPCs for the same volume in parallel - which, in our case, doesn't help the situation at all and would only create more unnecessary load on Manila itself. To guard against situations where there's a long running operation, some form of locking needs to be employed.

Using a simple mutex per volume ID would result in this:
```
  R req1             req2             req3
T -----------------+----------------+----------------+
01| LOCK           |                |                |
02| op1            |                |                |
03| op2            | LOCK...wait    |                |
04| op3            | wait...        |                |
05| UNLOCK         | wait...        | LOCK...wait    |
06| RETURN success + LOCK           | wait...        |
07|                  op1            | wait...        |
08|                  op2            | wait...        |
09|                  op3            | wait...        |
21|                  UNLOCK         | wait...        |
22|                  RETURN success + LOCK           |
23|                                   op1            |
24|                                   op2            |
25|                                   op3            |
26|                                   UNLOCK         |
27|                                   RETURN SUCCESS +
```
_(I really hope these diagrams make at least a bit of sense)_ On the Y axis, there's `T` for time, and on the X axis there's `R` for a request. The first request `req1` locks the mutex and launches a long running operation `op` (taking up 3 time units). The call times out, so the CO decides to retry with `req2` but needs to wait till `req1` finishes so that `req2` may acquire the lock. This pattern may repeat until the call doesn't time-out. There are two problems here: (a) it may take a long time till CO is satisfied with the answer, (b) the long running operation `op` is being run in all of those retries.

`pkg/csi/manila/responsebroker` from this PR tries to tackle the problem (b), and hopes to mitigate (a):
```
  R req1                                             req2                          req3
T -------------------------------------------------+-----------------------------+-----------------------------+
01| LOCK(acq-or-create)                            |                             |                             |
02| op1                                            |                             |                             |
03| op2                                            | LOCK(ack-or-create)...wait  |                             |
04| op3                                            | wait...                     |                             |
05| UNLOCK, write success, wait till all Rs finish | wait...                     | LOCK(ack-or-create)...wait  |
06| wait...                                        | LOCK, read response:success | wait...                     |
07| wait...                                        | UNLOCK                      | wait...                     |
08| wait...                                        | RETURN SUCCESS              + LOCK, read response:success |
09| RETURN SUCCESS                                 +                               UNLOCK                      |
10|                                                                                RETURN SUCCESS              +
```
The same scenario as above: `req1` locks the mutex and runs the long running operation `op` and times-out. `req2` will wait till it can acquire the lock, but instead of running `op` again, it reads the response from `req1` and if it's successful, it immediately exits with success. The `responsebroker` basically caches the result of an operation and propagates it to multiple readers. 

This mechanism could be also used to implement a form of journaling: if an error occurs in one call, the future retries would pick up the operation in the exact spot where the previous call had failed - again, saving some load from the backend. Maybe in another PR.

### To be done in this PR

* If an operator wishes to use a custom certificate for authentication, they may use the `os-certAuthority` parameter in Secrets to inject the contents of a PEM cert file. This is the way manila-provisioner does it. The problem here is the size limit: the [CSI spec allows up to 128 bytes](https://github.com/container-storage-interface/spec/blob/master/spec.md#size-limits) in `string` fields and a certificate can be obviously larger than that,
* `List*` and `GetCapacity` CSI RPCs need OpenStack credentials, but these RPCs don't carry any Secrets fields. CSI Cinder driver solves this by supplying the driver with a file containing the credentials. It doesn't rely on Secrets supplied by the CO.

**Release note**:
```release-note
added CSI Manila driver
```

/cc @tsmetana @dims @adisky

adisky commented:
/ok-to-test

tsmetana commented:
Hi. Overall I'd say the design looks OK. I would prefer to have some more tests to accompany the code. Otherwise it looks like a good start.

gouthampacha commented on pkg/csi/manila/options/shareoptions.go:
```diff
@@ -0,0 +1,65 @@
+	// Adapter options
+
+	CephfsMounter  string `name:"cephfs-mounter" value:"default:fuse" matches:"^kernel|fuse$"`
+	NFSShareClient string `name:"nfs-shareClient" value:"default:0.0.0.0"`
```
0.0.0.0/0

tsmetana replied:
There were issues with Ganesha NFS server, see e.g.:
https://bugs.launchpad.net/manila/+bug/1800627

And for some reason 0.0.0.0 worked. So this is where this oddity comes from. If this is not the case any longer then I agree the /0 should be there.

tombarron replied:
This ganesha oddity was https://bugs.launchpad.net/manila/+bug/1800627 which we addressed by having the manila driver accept the conventional 0.0.0.0/0 but present (and recognize) 0.0.0.0 without the slash-zero when interacting with the ganesha service itself.  Fix was released and backported.

gman0 replied:
Will do + i'll try to write some unit tests, probably next week or so though...completely swamped with school...

gman0 replied:
Sorry for the hold up, one more busy week

gman0 replied:
Done in https://github.com/kubernetes/cloud-provider-openstack/pull/536/commits/73bea65a8f7a0295980b526b7089b0c24c8badf3

tombarron commented on cmd/manila-csi-plugin/main.go:
```diff
@@ -0,0 +1,53 @@
+	endpoint    = flag.String("endpoint", "unix://tmp/csi.sock", "CSI endpoint")
+	driverName  = flag.String("drivername", "manila.csi.openstack.org", "name of the driver")
+	nodeID      = flag.String("nodeid", "", "node id")
+	shareKind   = flag.String("sharekind", "", "kind of the Manila share along with the CSI endpoint for a dedicated CSI node plugin (expects a value in the format of SHARE-PROTO=UNIX-DOMAIN-SOCKET)")
```
I wonder if we can think about a more specific name for this variable than ``shareKind``, maybe something along the lines of ``shareProtocolSelector``?  There are various ways of making "kinds" of shares in Manila so the very general name might confuse.

gman0 replied:
Definitely! Accurate terminology is important, this one was just a wip name

gman0 replied:
Done in https://github.com/kubernetes/cloud-provider-openstack/pull/536/commits/a773efe7cd37c552b85337124d07c46445939c2b

tombarron commented on pkg/csi/manila/options/shareoptions.go:
```diff
@@ -28,7 +28,7 @@ type ControllerVolumeContext struct {
 
 	CephfsMounter  string `name:"cephfs-mounter" value:"default:fuse" matches:"^kernel|fuse$"`
-	NFSShareClient string `name:"nfs-shareClient" value:"default:0.0.0.0"`
+	NFSShareClient string `name:"nfs-shareClient" value:"default:0.0.0.0/0"`
```
Thanks!

tombarron commented on cmd/manila-csi-plugin/main.go:
```diff
@@ -18,18 +18,16 @@ package main
+	endpoint      = flag.String("endpoint", "unix://tmp/csi.sock", "CSI endpoint")
+	driverName    = flag.String("drivername", "manila.csi.openstack.org", "name of the driver")
+	nodeID        = flag.String("nodeid", "", "node id")
+	protoSelector = flag.String("share-protocol-selector", "", "specifies which Manila share protocol to use")
```
Cool, thanks.

tsmetana commented on pkg/csi/manila/controllerserver.go:
```diff
@@ -0,0 +1,228 @@
+	}
+
+	if share.Size != sizeInGiB {
+		return nil, status.Errorf(codes.AlreadyExists, "volume %s already exists, but is incompatible with the request")
```
There is an argument missing (volID most probably).

tsmetana commented on pkg/csi/manila/nodeserver.go:
```diff
@@ -0,0 +1,329 @@
+	if strings.ToLower(share.ShareProto) != strings.ToLower(ns.d.shareProto) {
+		return nil, nil, status.Errorf(codes.InvalidArgument,
+			"wrong share protocol %s for volume %s (share ID %s), the plugin is set to operate in %s",
+			share.ShareProto, share.ID, ns.d.shareProto)
```
Again: missing an argument (volID), the string format expects one more.

tsmetana commented on pkg/csi/manila/nodeserver.go:
```diff
@@ -0,0 +1,329 @@
+		if ok {
+			volumeCtx, secrets = cacheEntry.volumeContext, cacheEntry.secrets
+		} else {
+			klog.Errorf("STAGE_UNSTAGE_VOLUME capability is enabled, but node stage cache doesn't contain an entry for %s - this is most likely a bug! Rebuild staging data anyway...")
```
And once more: missing argument for the string format.

gman0 replied:
Thanks for noticing those!

gman0 replied:
Fixed in https://github.com/kubernetes/cloud-provider-openstack/pull/536/commits/91be405578e36c59d7f69f1c6afdbb364348989f

gman0 commented on examples/manila-csi-plugin/README.md:
```diff
@@ -2,8 +2,15 @@
+3. Modify `secrets.yaml` to suite your OpenStack cloud environment. Refer to the _"Secrets, authentication"_ section of CSI Manila docs. You may also use helper scripts from `examples/manila-provisioner` to generate the Secrets manifest.
+4. Deploy OpenStack secrets
+5. Create a persistent volume:
+  5.1 **If you want to provision a new share:**
```
Nested numbered lists are not rendered correctly

gman0 commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.14

gman0 commented:
I can't seem to make cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.14 and cloud-provider-openstack-acceptance-test-e2e-conformance tests succeed. The logs say

```
19-05-04 09:23:06.403840 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:06 process.go:153: Running: make -C /home/zuul/src/k8s.io/kubernetes quick-release
2019-05-04 09:23:06.409060 | ubuntu-xenial-vexxhost | make: Entering directory '/home/zuul/src/k8s.io/kubernetes'
2019-05-04 09:23:06.468435 | ubuntu-xenial-vexxhost | /home/zuul/src/k8s.io/kubernetes/hack/lib/golang.sh: line 295: KUBE_SERVER_PLATFORMS[@]: unbound variable
2019-05-04 09:23:06.468545 | ubuntu-xenial-vexxhost | make: *** [quick-release] Error 1
2019-05-04 09:23:06.468661 | ubuntu-xenial-vexxhost | Makefile:460: recipe for target 'quick-release' failed
2019-05-04 09:23:06.468724 | ubuntu-xenial-vexxhost | make: Leaving directory '/home/zuul/src/k8s.io/kubernetes'
2019-05-04 09:23:06.475472 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:06 process.go:155: Step 'make -C /home/zuul/src/k8s.io/kubernetes quick-release' finished in 71.961665ms
2019-05-04 09:23:06.475979 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:06 process.go:96: Saved XML output to /home/zuul/workspace/logs/kubernetes/junit_runner.xml.
2019-05-04 09:23:06.476529 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:06 process.go:153: Running: bash -c . hack/lib/version.sh && KUBE_ROOT=. kube::version::get_version_vars && echo "${KUBE_GIT_VERSION-}"
2019-05-04 09:23:07.158458 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:07 process.go:155: Step 'bash -c . hack/lib/version.sh && KUBE_ROOT=. kube::version::get_version_vars && echo "${KUBE_GIT_VERSION-}"' finished in 681.300364ms
2019-05-04 09:23:07.158740 | ubuntu-xenial-vexxhost | 2019/05/04 09:23:07 main.go:314: Something went wrong: failed to acquire k8s binaries: error during make -C /home/zuul/src/k8s.io/kubernetes quick-release: exit status 2
```

Seems like the problem is with the tests themselves?

ramineni commented:
@gman0 tests seem to fail on all patch for quite some days now :( .. should be fixed k8s main repo I suppose. we are facing issue mentioned https://github.com/kubernetes/kubernetes/issues/77521. hopefully would be resolved soon.

gman0 commented:
@ramineni thanks for the info!

gman0 commented on examples/manila-csi-plugin/preprovisioned-pvc.yaml:
```diff
@@ -0,0 +1,37 @@
+      csi.storage.k8s.io/node-stage-secret-name: csi-manila-secrets
+      csi.storage.k8s.io/node-stage-secret-namespace: default
+      csi.storage.k8s.io/node-publish-secret-name: csi-manila-secrets
+      csi.storage.k8s.io/node-publish-secret-namespace: default
```
Secrets must be passed directly via SecretsRef

gman0 commented:
If nobody opposes, i believe this PR is ready. IMHO this is a reasonable starting point and we can continue with improvements in the upcoming PRs.

As a part of technical debt for this project, unit and E2E tests are still missing (TBD during this year's Google Summer of Code in which I'm participating). The same goes for Availability Zones support: this is planned after GSoC, though contributions are of course very much welcome if anybody would like to pick this up.

@tsmetana @tombarron @gouthampacha @dims @adisky if you guys don't have any further comments i can squash the commits

adisky commented:
@gman0 I have no problem, Improvements can be done later. Just one thing the, you can add its design description as doc or readme , so it will be helpful for other contributors

gman0 commented:
@adisky i've added dev documentation under `docs/developers-csi-manila.md`: https://github.com/kubernetes/cloud-provider-openstack/blob/00a62387f3485df614b890080b70e7519321723c/docs/developers-csi-manila.md

the second half of the document is just copied over from this PR's description - hope that's ok!

Other than that, NFS shares should be now working as well and i've split `ShareAdapter`'s `BuildSecrets()` into `BuildNodeStageSecret()` and `BuildNodePublishSecret()` in order to be more CSI compliant.

gman0 commented:
/test cloud-provider-openstack-acceptance-test-standalone-cinder

gman0 commented:
/test cloud-provider-openstack-acceptance-test-flexvolume-cinder

gman0 commented:
cloud-provider-openstack-acceptance-test-flexvolume-cinder:
```
2019-05-17 07:08:58.710022 | ubuntu-xenial | Err:17 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libvirt0 amd64 1.3.1-1ubuntu10.25
2019-05-17 07:08:58.710130 | ubuntu-xenial |   404  Not Found [IP: 91.189.88.149 80]
2019-05-17 07:08:58.765422 | ubuntu-xenial | Ign:18 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libvirt-bin amd64 1.3.1-1ubuntu10.25
2019-05-17 07:08:58.949240 | ubuntu-xenial | Err:17 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libvirt0 amd64 1.3.1-1ubuntu10.25
2019-05-17 07:08:58.949346 | ubuntu-xenial |   404  Not Found [IP: 91.189.88.149 80]
2019-05-17 07:08:59.057629 | ubuntu-xenial | Err:18 http://security.ubuntu.com/ubuntu xenial-security/main amd64 libvirt-bin amd64 1.3.1-1ubuntu10.25
2019-05-17 07:08:59.057736 | ubuntu-xenial |   404  Not Found [IP: 91.189.88.149 80]
2019-05-17 07:08:59.062074 | ubuntu-xenial | Fetched 1,294 kB in 2s (559 kB/s)
2019-05-17 07:08:59.062311 | ubuntu-xenial | E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/libv/libvirt/libvirt0_1.3.1-1ubuntu10.25_amd64.deb  404  Not Found [IP: 91.189.88.149 80]
2019-05-17 07:08:59.062347 | ubuntu-xenial |
2019-05-17 07:08:59.062544 | ubuntu-xenial | E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/libv/libvirt/libvirt-bin_1.3.1-1ubuntu10.25_amd64.deb  404  Not Found [IP: 91.189.88.149 80]
```
not much I can do about this

tsmetana commented:
I'm also quite unsure what to do with the latest test failures. As I mentioned long ago: I'm for merging this.

ramineni commented:
/test cloud-provider-openstack-acceptance-test-flexvolume-cinder

gman0 commented:
Commits are squashed, too bad about the tests still failing...anyway, if there are no more comments I believe this PR to be done.

adisky commented:
@gman0 we are facing gate failure from pas few days, trying to get is resolved asap.

gman0 commented:
@adisky ok no probs, thanks for the heads up!

mrhillsman commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance

mrhillsman commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.12

mrhillsman commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.13

mrhillsman commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance-stable-branch-v1.14

mrhillsman commented:
/test cloud-provider-openstack-test-flexvolume-cinder

mrhillsman commented:
/test cloud-provider-openstack-acceptance-test-flexvolume-cinder

gman0 commented:
/test cloud-provider-openstack-acceptance-test-e2e-conformance

gman0 commented:
Tests are finally passing! @dims @hogepodge @adisky PTAL

adisky commented:
@gman0 Thanks!!!
/lgtm

adisky commented:
/priority important-longterm

tsmetana commented:
@dims or @rootfs Could you please take a look and approve the merge eventually? Thanks.

dims commented:
/approve

gman0 commented:
@adisky @dims Thanks guys! It seems like k8s-ci-robot wasn't triggered though..?

adisky commented:
@dims it has not taken "/approved" label, you need to do it again
